src/lua/version.lua

```diff
@@ -0,0 +1,248 @@
+        minor = minor,
+        patch = patch,
+        _prerelease_name_weight = prerelease_weight,
+        _prerelease_num = prerelease_num,
+        commit = commit,
+    })
+end
+
+local function version_fromstr(version_str)
+    check_param(version_str, 'version_str', 'string')
+    --  x.x.x-name<num>-<num>-g<commit>
+    -- \____/\________/\_____/
+    --   P1      P2      P3
+    local major, minor, patch, prerelease, commit, pos
+
+    -- Part 1 - version ID triplet. Compulsory fields.
+    major, minor, patch = version_str:match('^(%d+)%.(%d+)%.(%d+)')
+    if not major or not minor or not patch then
+        local msg = "Error during parsing version string '%s'. " ..
+                    "At least X.Y.Z is required."
+        box.error(box.error.ILLEGAL_PARAMS, string.format(msg, version_str))
+    end
+    major = tonumber(major)
+    minor = tonumber(minor)
+    patch = tonumber(patch)
+    -- Cut to 'type<num>-<num>-g<commit>'.
+    pos = version_str:find('-')
+    if not pos then
+        goto finish
+    end
+    version_str = version_str:sub(pos + 1)
+
+    -- Part 2 - prerelease name and number, might be absent.
+    prerelease = version_str:match('^(%a+%d*)')
+    -- Cut to '<num>-g<commit>'.
+    if prerelease then
+        pos = version_str:find('-')
+        if not pos then
+            goto finish
+        end
+        version_str = version_str:sub(pos + 1)
+    end
+
+    -- Part 3 - commit count since the latest release, might be absent.
+    commit = version_str:match('^(%d+)')
+    if commit then
+        commit = tonumber(commit)
+    end
+
+::finish::
+    return version_new(major, minor, patch, prerelease, commit)
+end
+
+local function version_cmp(ver1, ver2)
+    if ver1.major ~= ver2.major then
+        return ver1.major - ver2.major
+    end
+    if ver1.minor ~= ver2.minor then
+        return ver1.minor - ver2.minor
+    end
+    if ver1.patch ~= ver2.patch then
+        return ver1.patch - ver2.patch
+    end
+    if ver1._prerelease_name ~= ver2._prerelease_name then
+        return ver1._prerelease_name_weight - ver2._prerelease_name_weight
+    end
+    if ver1._prerelease_num ~= ver2._prerelease_num then
+        return ver1._prerelease_num - ver2._prerelease_num
+    end
+    if ver1.commit ~= ver2.commit then
+        return ver1.commit - ver2.commit
+    end
+    return 0
+end
+
+local version_ctype
+local function version_cast_if_needed(version)
+    if type(version) == 'cdata' then
+        if ffi.typeof(version) ~= version_ctype then
+            return nil
+        end
+        -- Nothing to do.
+        return version
+    end
+    if type(version) == 'string' then
+        -- Convert to version object.
+        return version_fromstr(version)
+    end
+    return nil
+end
+
+local version_mt = {
+    __index = function(v, field)
+        if field == 'prerelease' then
+            return v._prerelease_name .. v._prerelease_num
+        elseif field == '_prerelease_name' then
```

Comment:
Do we really need to export `_prerelease_name` here? Users should only use `prerelease` while for comparisons and internal operations `_prerelease_id` and `_prerelease_num` should be enough.

Reply:
Nah, not really, we can make it a function, it was just convenient to use